### ISSUE TITLE
Make LibGit2Sharp aware of external changes to the index

### ISSUE THREAD

**Issue by nulltoken** [libgit2/libgit2sharp#526]
With #525 a new method `ReloadFromDisk()` has been added to the `Index` which overwrites its content with the state of the on-disk index file.

`RetrieveStatus` already calls this method.

This method call should be deployed to many more methods: most of the Index methods along with some other ones (`Commit`, `Stash`, `Reset`, ...).

Maybe the first step of this change would be to try to list all the candidates...

**Comment by ethomson** [comment]
I agree that this would be useful.  Looks like there's no API change here (unless we want to make this an optional thing for `Stash`).  Dropping the API stabilization tag.
